### PR TITLE
Toolbar overflow

### PR DESCRIPTION
**Pull request by TomTirapani**
Shows overflowing toolbar items in a dropdown. #1524

Screenshot 2020-02-26 at 12 32 28

Note that this does not work for vertical toolbars, due to how Blueprint detects overflow requirements by testing flex-shrink on width only: https://github.com/palantir/blueprint/blob/fddba2c6/packages/core/src/components/overflow-list/overflowList.tsx#L237

Also note that as a side-effect of this change, all toolbar items must be direct children of the toolbar. If children are nested, the entire outer component will hide. This is particularly apparent for toolbar forms, where you may have a form wrapping the toolbar items.

Please see corresponding Toolbox branch: https://github.com/xh/toolbox/pull/369

Hoist P/R Checklist
-------------------

**Pull request authors:** Review and check off the below. Items that do not apply can also be
checked off to indicate they have been considered. If unclear if a step is relevant, please leave
unchecked and note in comments.

- [x] Caught up with `develop` branch as of last change.
- [x] Added CHANGELOG entry, or determined not required.
- [x] Reviewed for breaking changes, added `breaking-change` label + CHANGELOG if so.
- [x] Updated doc comments / prop-types, or determined not required.
- [x] Reviewed and tested on Mobile, or determined not required.
- [x] Created Toolbox branch / PR, or determined not required.

**If your change is still a WIP**, please use the "Create draft pull request" option in the split
button below to indicate it is not ready yet for a final review.

> **Pull request reviewers:** when merging this P/R, please consider using a *squash commit* to
> collapse multiple intermediate commits into a single commit representing the overall feature
> change. This helps keep the commit log clean and easy to scan across releases. PRs containing a
> single commit should be *rebased* when possible.

### CODE REVIEW

File: desktop/cmp/toolbar/Toolbar.js [hunk 1]
```diff
@@ -4,36 +4,91 @@
...
     render(props) {
-        const {vertical, className, ...rest} = props;
+        const {children, className, vertical, collapseFrom = 'end', minVisibleItems, ...rest} = props;
 
         return (vertical ? vbox : hbox)({
             ...rest,
-            className: classNames(className, vertical ? 'xh-toolbar--vertical' : null)
+            className: classNames(className, vertical ? 'xh-toolbar--vertical' : null),
+            item: itemContainer({
+                $items: castArray(children),
```
Comment (lbwexler) [inline desktop/cmp/toolbar/Toolbar.js]: nice to see our dollar sign escape hatch in action!

File: desktop/cmp/toolbar/Toolbar.js [hunk 2]
```diff
@@ -4,36 +4,91 @@
...
         });
     }
 });
 
+const itemContainer = hoistCmp.factory({
+    render({items, vertical, minVisibleItems, collapseFrom}) {
+        if (!items || !items.length) return null;
+
+        // Overflow collapse is not supported for vertical toolbars
+        if (vertical) return fragment(items);
```
Comment (lbwexler) [inline desktop/cmp/toolbar/Toolbar.js]: clever!

File: desktop/cmp/toolbar/Toolbar.js [hunk 3]
```diff
@@ -4,36 +4,91 @@
...
+
+        return overflowList({
+            $items: items,
+            minVisibleItems,
```
Comment (lbwexler) [inline desktop/cmp/toolbar/Toolbar.js]: might be able to get the key's onto these items by using React.Children.toArray

File: desktop/cmp/toolbar/Toolbar.js [hunk 4]
```diff
@@ -4,36 +4,91 @@
...
+            },
+            overflowRenderer: (overflowItems) => {
+                return itemOverflowButton({overflowItems});
+            }
+        });
+    }
+});
+
+const itemOverflowButton = hoistCmp.factory({
+    render({overflowItems}) {
```
Comment (lbwexler) [inline desktop/cmp/toolbar/Toolbar.js]: again, would it be simpler just to pass these items as children to this button?

File: CHANGELOG.md [hunk 5]
```diff
@@ -4,6 +4,7 @@
 
 ### 🎁 New Features
 
+* Horizontal toolbars will collapse overflowing items into a drop down menu.
```
Comment (lbwexler) [inline CHANGELOG.md]: Think we need to highlight the breaking change, that toolbars now cannot be passed in container wrapping components

File: desktop/cmp/toolbar/Toolbar.js [hunk 6]
```diff
@@ -4,36 +4,91 @@
...
+
 Toolbar.propTypes = {
     /** Custom classes that will be applied to this component */
     className: PT.string,
 
     /** Set to true to vertically align the items of this toolbar */
-    vertical: PT.bool
+    vertical: PT.bool,
+
+    /** Which direction the items should collapse from: 'start' or 'end' of the children. */
```
Comment (lbwexler) [inline desktop/cmp/toolbar/Toolbar.js]: Mention again here that this is for Horizontal Toolbars only?  Mention defaults?

**Comment by lbwexler** [comment]
We have decided to make this opt-out with an additional flag on toolbar `enableOverflowMenu'.  Should default to true .

**Comment by lbwexler** [comment]
So, I added in the opt-out flag and cleaned up a few things.  But then realized that there is the weird situation, the labels will now wrap when opting out with enableOverflowMenu: false , where previously they would clip !?!.

Tom -- can we take a look together on monday?


Saw this, in e.g. the toolbar for the log viewer.

**Comment by TomTirapani** [comment]
I think that's to be expected - we also have to opt out of the CSS styling changes. I'll take a look.

**Comment by TomTirapani** [comment]
Hi @lbwexler - I've updated the PR to restore the no-wrap and default padding in non-overflow mode

File: CHANGELOG.md [hunk 7]
```diff
@@ -95,6 +96,8 @@
 
 ### 💥 Breaking Changes
 
+* `Toolbar` items must be provided as direct children. Wrapping Toolbar items in container
+  components can result in unexpected item overflow.
```
Comment (amcclain) [inline CHANGELOG.md]: Glad we got this in, but note these ended up under the wrong changelog version - fixing now.

**Pull request merged**